### PR TITLE
[engine] Remove kernel-level side effects push in refine.

### PR DESCRIPTION
**Pull request by ejgallego**
This was noted by Gaëtan after some investigations, and indeed it
makes sense refine should have no business messing with this.

The functions creating the side-effects should be the ones that ensure
that the env is consistent.

I'm not fully sure how this is working yet tho, this part of the proof
engine is hard to follow.

Co-authored-by: Gaëtan Gilbert <gaetan.gilbert@skyskimmer.net>

### CODE REVIEW

**Comment by gares** [comment]
```
File "./src/Rewriter/Util/Bool/Reflect.v", line 311, characters 9-117:
Error:
Anomaly "File "clib/int.ml", line 46, characters 14-20: Assertion failed."
Please report at http://coq.inria.fr/bugs/.
```'
I get these quite a lot in coq-elpi when I mess up the environments (eg. typecheck in environment e2 a term containing constants that live in e1 but not e2).

**Comment by ppedrot** [comment]
Likely related to the fact mentioned on Zulip, i.e. there is some thunk that performed side-effects that was not correctly replayed in this PR since it simply doesn't do anything. You have to wrap `Refine.lift` in the effect-adding layer (and not touch `Safe_typing.push_private_constants`).

**Comment by ejgallego** [comment]
Only failure seems in rewriting which has it's own tactics, could be possibly `build_constant_by_tactic` and such stuff, which I believe broken.

**Comment by ejgallego** [comment]
Likely culprit?
```ocaml
let make_definition_by_tactic sigma ~poly (name : Names.Id.t) (ty : EConstr.t) (tac : unit Proofview.tactic) =
  let cinfo = Declare.CInfo.make ~name ~typ:ty () in
  let info = Declare.Info.make ~poly () in
  let lemma = Declare.Proof.start ~cinfo ~info sigma in
  let lemma, _ = Declare.Proof.by tac lemma in
  let ids = Declare.Proof.save_regular ~proof:lemma ~opaque:Vernacexpr.Transparent ~idopt:None in
  match ids with
  | [Names.GlobRef.ConstRef cst] -> cst
  | _ -> CErrors.user_err (Pp.str "Internal error in make_definition_by_tactic")
```
https://github.com/mit-plv/rewriter/blob/master/src/Rewriter/Util/plugins/definition_by_tactic.ml.v813

Maybe that's used inside a tactic? I don't think so, but not easy to follow the context here https://github.com/mit-plv/rewriter/blob/master/src/Rewriter/Util/Bool/Reflect.v#L311

**Comment by gares** [comment]
What is wrong in the code above? It seems OK to me. Why it is the culprit?

**Review by SkySkimmer — CHANGES_REQUESTED** [review]
This is clearly incorrect, we got the env from before running the tactic so it cannot contain any new constants.
If we can find the original discussion again we may be able to contruct a better patch, otherwise we can only close the pr.
